lib/opencensus/common/config.rb

```diff
@@ -0,0 +1,370 @@
+      # navigate to the subconfig. Names may comprise only letters, numerals,
+      # and underscores, and must begin with a letter.
+      #
+      # If you provide a block, the subconfig object is passed to the block,
+      # so you can easily add fields to the subconfig.
+      #
+      # You may also pass in a config object that already exists. This will
+      # "attach" that configuration in this location.
+      #
+      # @param [String, Symbol] key The name of the subconfig
+      # @param [Config] config A config object to attach here. If not provided,
+      #     creates a new config.
+      #
+      # @return [Config] self for chaining
+      #
+      def add_config! key, config = nil, &block
+        key = validate_key! key
+        if config.nil?
+          config = Config.new(&block)
+        elsif block
+          yield config
+        end
+        @fields[key] = config
+        define_getter_method! key
+        self
+      end
+
+      ##
+      # Assign an option with the given name to the given value.
+      #
+      # @param [Symbol, String] key The option name
+      # @param [Object] value The new option value
+      #
+      def []= key, value
+        key = key.to_sym
+        unless @fields.key? key
+          raise ArgumentError, "Key #{key.inspect} does not exist"
+        end
+        field = @fields[key]
+        if field.is_a? Config
+          raise ArgumentError, "Key #{key.inspect} is a subconfig"
+        end
+        validate_value! field, value
+        field.value = value
+      end
+
+      ##
+      # Get the option or subconfig with the given name.
+      #
+      # @param [Symbol, String] key The option or subconfig name
+      # @return [Object] The option value or subconfig object
+      #
+      def [] key
+        key = key.to_sym
+        unless @fields.key? key
+          raise ArgumentError, "Key #{key.inspect} does not exist"
+        end
+        field = @fields[key]
+        if field.is_a? Config
+          field
+        else
+          field.value
+        end
+      end
+
+      ##
+      # Check if this Config object has an option of the given name.
+      #
+      # @param [Symbol] key The key to check for.
+      # @return [boolean] true if the inquired key is a valid option for this
+      #   Config object. False otherwise.
+      #
+      def option? key
+        @fields[key.to_sym].is_a? Option
+      end
+
+      ##
+      # Check if this Config object has a subconfig of the given name.
+      #
+      # @param [Symbol] key The key to check for.
+      # @return [boolean] true if the inquired key is a valid subconfig of this
+      #   Config object. False otherwise.
+      #
+      def subconfig? key
+        @fields[key.to_sym].is_a? Config
+      end
+
+      ##
+      # Return a list of valid option names.
+      #
+      # @return [Array<Symbol>] a list of option names as symbols.
+      #
+      def options!
```

Comment:
Why are some of these bang(!) methods? It doesn't seem to raise an exception or modify the current object.

Same goes for `subconfigs!`, `to_s!`, `to_h!`

Reply:
Yeah... that was a design decision, to make _all_ the methods end with ! or ? to avoid polluting the namespace (any more than Object and Kernel already do, and then I tried to avoid calling Object/Kernel methods unqualified in case they do get overridden). I even experimented with extending BasicObject, but that causes other headaches so I dropped it. This is all open to discussion of course.